head.go

```diff
@@ -601,9 +601,112 @@ func (h *Head) Delete(mint, maxt int64, ms ...labels.Matcher) error {
 	for _, s := range stones {
 		h.tombstones.add(s.ref, s.intervals[0])
 	}
+
+	_, err = h.CleanTombstones()
+	if err != nil {
+		return errors.Wrap(err, "Head: CleanTombstones in Delete")
+	}
+
 	return nil
 }
 
+// CleanTombstones re-writes the chunks with tombstones
+// and deletes chunks if there is no samples after removing tombstones 
+func (h *Head) CleanTombstones() (bool, error) {
+
+	modified := false // flag for entire block
+
+	for sid, dranges := range h.tombstones {
+		if len(dranges) == 0 {
+			continue
+		}
+
+		cleaned := false // flag for current tombstone
+		stripeSeriesIndex := sid%stripeSize
+
+		h.series.locks[stripeSeriesIndex].Lock()
+		s := h.series.series[stripeSeriesIndex] // series with tombstones
+
+		if ms, ok := s[sid]; ok {
```

Comment:
Need to lock `ms` too.

Reply:
Line 627 locks the series, and `ms` is obtained from that series. So why lock `ms` again?